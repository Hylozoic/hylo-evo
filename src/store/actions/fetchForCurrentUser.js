import { get } from 'lodash/fp'
import { FETCH_FOR_CURRENT_USER } from 'store/constants'
import communityTopicsQueryFragment from 'graphql/fragments/communityTopicsQueryFragment'
import communityQueryFragment from 'graphql/fragments/communityQueryFragment'
import meQueryFragment from 'graphql/fragments/meQueryFragment'

export default function (slug, skipTopics) {
  const query = slug
    ? `query ($slug: String, $first: Int, $offset: Int, $sortBy: String, $order: String, $autocomplete: String, $subscribed: Boolean, $updateLastViewed: Boolean) {
      ${meQueryFragment}
      ${communityQueryFragment}
    }`
    : (skipTopics
      ? `{
        ${meQueryFragment}
      }`
      : `query ($first: Int, $offset: Int, $sortBy: String, $order: String, $autocomplete: String, $subscribed: Boolean) {
        ${meQueryFragment}
        ${communityTopicsQueryFragment}
      }`)

  return {
    type: FETCH_FOR_CURRENT_USER,
    graphql: {query, variables: queryVariables(slug)},
    meta: {
      extractModel: [
        {
          getRoot: get('me'),
          modelName: 'Me',
          append: true
        },
        {
          getRoot: get(slug ? 'community' : 'communityTopics'),
          modelName: slug ? 'Community' : 'CommunityTopic',
          append: true
        }
      ]
    }
  }
}

// the value of `first` is high because we are receiving unaggregated data from
// the API, so there could be many duplicates
const queryVariables = slug => ({slug, first: 200, offset: 0, subscribed: true, updateLastViewed: true})
