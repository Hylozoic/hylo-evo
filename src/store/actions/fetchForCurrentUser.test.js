import fetchForCurrentUser from './fetchForCurrentUser'

it('returns only current user fetch query when there is not any slug and skipTopics is true', () => {
  expect(fetchForCurrentUser(null, true)).toMatchSnapshot()
})

it('doesn\'t return topics if there is a slug of skipTopics is true', () => {
  expect(fetchForCurrentUser('anyslug', true)).toMatchSnapshot()
})

it('doesn\'t return topics if there is a slug of skipTopics is false', () => {
  expect(fetchForCurrentUser('anyslug', false)).toMatchSnapshot()
})
