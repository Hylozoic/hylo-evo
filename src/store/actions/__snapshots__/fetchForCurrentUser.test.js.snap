// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't return topics if there is a slug of skipTopics is false 1`] = `
Object {
  "graphql": Object {
    "query": "query ($slug: String, $first: Int, $offset: Int, $sortBy: String, $order: String, $autocomplete: String, $subscribed: Boolean, $updateLastViewed: Boolean) {
      
me {
  id
  name
  avatarUrl
  newNotificationCount
  unseenThreadCount
  location
  email
  settings {
    signupInProgress
  }
  memberships {
    id
    lastViewedAt
    newPostCount
    hasModeratorRole
    community {
      id
      name
      slug
      avatarUrl
      network {
        id
        slug
        name
        avatarUrl
        communities(first: 100) {
          items {
            id
          }
        }
      }
    }
  }
}
      
community(slug: $slug, updateLastViewed: $updateLastViewed) {
  id
  name
  slug
  description
  avatarUrl
  network {
    id
    slug
    name
    avatarUrl
    communities(first: 100) {
      items {
        id
      }
    }
  }
  memberCount
  members(first: 8, sortBy: \\"name\\", order: \\"desc\\") {
    items {
      id
      name
      avatarUrl
    }
  }
  moderators {
    items {
      id
      name
      avatarUrl
    }
  }
  
communityTopics(
  first: $first,
  offset: $offset,
  sortBy: $sortBy,
  order: $order,
  subscribed: $subscribed,
  autocomplete: $autocomplete
) {
  hasMore
  total
  items {
    id
    postsTotal
    followersTotal
    isSubscribed
    newPostCount
    topic {
      id
      name
    }
  }
}

}
    }",
    "variables": Object {
      "first": 200,
      "offset": 0,
      "slug": "anyslug",
      "subscribed": true,
      "updateLastViewed": true,
    },
  },
  "meta": Object {
    "extractModel": Array [
      Object {
        "append": true,
        "getRoot": [Function],
        "modelName": "Me",
      },
      Object {
        "append": true,
        "getRoot": [Function],
        "modelName": "Community",
      },
    ],
  },
  "type": "FETCH_FOR_CURRENT_USER",
}
`;

exports[`doesn't return topics if there is a slug of skipTopics is true 1`] = `
Object {
  "graphql": Object {
    "query": "query ($slug: String, $first: Int, $offset: Int, $sortBy: String, $order: String, $autocomplete: String, $subscribed: Boolean, $updateLastViewed: Boolean) {
      
me {
  id
  name
  avatarUrl
  newNotificationCount
  unseenThreadCount
  location
  email
  settings {
    signupInProgress
  }
  memberships {
    id
    lastViewedAt
    newPostCount
    hasModeratorRole
    community {
      id
      name
      slug
      avatarUrl
      network {
        id
        slug
        name
        avatarUrl
        communities(first: 100) {
          items {
            id
          }
        }
      }
    }
  }
}
      
community(slug: $slug, updateLastViewed: $updateLastViewed) {
  id
  name
  slug
  description
  avatarUrl
  network {
    id
    slug
    name
    avatarUrl
    communities(first: 100) {
      items {
        id
      }
    }
  }
  memberCount
  members(first: 8, sortBy: \\"name\\", order: \\"desc\\") {
    items {
      id
      name
      avatarUrl
    }
  }
  moderators {
    items {
      id
      name
      avatarUrl
    }
  }
  
communityTopics(
  first: $first,
  offset: $offset,
  sortBy: $sortBy,
  order: $order,
  subscribed: $subscribed,
  autocomplete: $autocomplete
) {
  hasMore
  total
  items {
    id
    postsTotal
    followersTotal
    isSubscribed
    newPostCount
    topic {
      id
      name
    }
  }
}

}
    }",
    "variables": Object {
      "first": 200,
      "offset": 0,
      "slug": "anyslug",
      "subscribed": true,
      "updateLastViewed": true,
    },
  },
  "meta": Object {
    "extractModel": Array [
      Object {
        "append": true,
        "getRoot": [Function],
        "modelName": "Me",
      },
      Object {
        "append": true,
        "getRoot": [Function],
        "modelName": "Community",
      },
    ],
  },
  "type": "FETCH_FOR_CURRENT_USER",
}
`;

exports[`returns only current user fetch query when there is not any slug and skipTopics is true 1`] = `
Object {
  "graphql": Object {
    "query": "{
        
me {
  id
  name
  avatarUrl
  newNotificationCount
  unseenThreadCount
  location
  email
  settings {
    signupInProgress
  }
  memberships {
    id
    lastViewedAt
    newPostCount
    hasModeratorRole
    community {
      id
      name
      slug
      avatarUrl
      network {
        id
        slug
        name
        avatarUrl
        communities(first: 100) {
          items {
            id
          }
        }
      }
    }
  }
}
      }",
    "variables": Object {
      "first": 200,
      "offset": 0,
      "slug": null,
      "subscribed": true,
      "updateLastViewed": true,
    },
  },
  "meta": Object {
    "extractModel": Array [
      Object {
        "append": true,
        "getRoot": [Function],
        "modelName": "Me",
      },
      Object {
        "append": true,
        "getRoot": [Function],
        "modelName": "CommunityTopic",
      },
    ],
  },
  "type": "FETCH_FOR_CURRENT_USER",
}
`;
